ext/random/random.c

```diff
@@ -0,0 +1,1754 @@
+	zval tmp;
+	int i;
+
+	for (i = 0; i< MT_N; i++) {
+		ZVAL_LONG(&tmp, s->s[i]);
+		zend_hash_next_index_insert(data, &tmp);
+	}
+	ZVAL_LONG(&tmp, s->cnt);
+	zend_hash_next_index_insert(data, &tmp);
+	ZVAL_LONG(&tmp, s->seeded);
+	zend_hash_next_index_insert(data, &tmp);
+
+	return SUCCESS;
+}
+
+static int mersennetwister_unserialize(void *state, HashTable *data) {
+	php_random_engine_state_mersennetwister *s = (php_random_engine_state_mersennetwister *)state;
+	zval *tmp;
+	int i;
+
+	for (i = 0; i < MT_N; i++) {
+		tmp = zend_hash_index_find(data, i);
+		if (!tmp || Z_TYPE_P(tmp) != IS_LONG) {
+			return FAILURE;
+		}
+
+		s->s[i] = Z_LVAL_P(tmp);
+	}
+	tmp = zend_hash_index_find(data, MT_N); /* cnt */
+	if (!tmp || Z_TYPE_P(tmp) != IS_LONG) {
+		return FAILURE;
+	}
+	s->cnt = Z_LVAL_P(tmp);
+	tmp = zend_hash_index_find(data, MT_N + 1); /* seeded */
+	if (!tmp || Z_TYPE_P(tmp) != IS_LONG) {
+		return FAILURE;
+	}
+	s->seeded = Z_LVAL_P(tmp);
+
+	return SUCCESS;
+}
+
+static zend_object *php_random_engine_mersennetwister_new(zend_class_entry *ce) {
+	return php_random_engine_common_init(ce, &php_random_engine_algo_mersennetwister, &random_engine_mersennetwister_object_handlers);
+}
+
+const php_random_engine_algo php_random_engine_algo_mersennetwister = {
+	sizeof(uint32_t),
+	mersennetwister_dynamic_generate_size,
+	sizeof(php_random_engine_state_mersennetwister),
+	mersennetwister_generate,
+	mersennetwister_seed,
+	mersennetwister_serialize,
+	mersennetwister_unserialize
+};
+
+/* MersenneTwister end */
+
+/* Secure begin */
+
+static inline size_t secure_dynamic_generate_size(void *state) {
+	return sizeof(uint64_t);
+}
+
+static uint64_t secure_generate(void *state) {
+	uint64_t ret;
+
+	php_random_bytes_silent(&ret, sizeof(uint64_t));
+
+	return ret;
+}
+
+static zend_object *php_random_engine_secure_new(zend_class_entry *ce) {
+	return php_random_engine_common_init(ce, &php_random_engine_algo_secure, &random_engine_secure_object_handlers);
+}
+
+const php_random_engine_algo php_random_engine_algo_secure = {
+	sizeof(uint64_t),
+	secure_dynamic_generate_size,
+	0,
+	secure_generate,
+	NULL,
+	NULL,
+	NULL
+};
+
+/* Secure end */
+
+/* User begin */
+
+static inline size_t user_dynamic_generate_size(void *state) {
+	php_random_engine_state_user *s = (php_random_engine_state_user *) state;
+
+	return s->last_generate_size;
+}
+
+static uint64_t user_generate(void *state) {
+	php_random_engine_state_user *s = (php_random_engine_state_user *) state;
+	uint64_t result = 0;
+	size_t size;
+	zval retval;
+	int i;
+	char *ptr;
+
+	zend_call_known_instance_method_with_0_params(s->generate_method, s->object, &retval);
+
+	/* Store generated size in a state */
+	size = Z_STR(retval)->len;
+	s->last_generate_size = size;
+
+	if (size > 0) {
+		/* Endianness safe copy */
+		ptr = (char *) &result;
+		for (i = 0; i < size; i++) {
+			ptr[i] = Z_STR(retval)->val[i];
+		}
```

Comment:
This loop is unsafe if the returned size is larger than the size of the `uint64_t`.

Reply:
fixed: c26b9c3